pkg/gpu/nvidia/metrics/metrics.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/NVIDIA/gpu-monitoring-tools/bindings/go/nvml"
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var (
+	// DutyCycle reports the percent of time when the GPU was actively processing.
+	DutyCycle = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "duty_cycle",
+			Help: "Percent of time when the GPU was actively processing",
+		},
+		[]string{"namespace", "pod", "container", "make", "accelerator_id", "model"})
+
+	// MemoryTotal reports the total memory available on the GPU.
+	MemoryTotal = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "memory_total",
+			Help: "Total memory available on the GPU in bytes",
+		},
+		[]string{"namespace", "pod", "container", "make", "accelerator_id", "model"})
+
+	// MemoryUsed reports GPU memory allocated.
+	MemoryUsed = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "memory_used",
+			Help: "Allocated GPU memory in bytes",
+		},
+		[]string{"namespace", "pod", "container", "make", "accelerator_id", "model"})
+
+	// AcceleratorRequests reports the number of GPU devices requested by the container.
+	AcceleratorRequests = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "accelerator_requests",
+			Help: "Number of accelerator devices requested by the container",
+		},
+		[]string{"namespace", "pod", "container", "resource_name"})
+)
+
+// MetricServer exposes GPU metrics for all containers in prometheus format on the specified port.
+type MetricServer struct {
+	collectionInterval  int
+	port                int
+	metricsEndpointPath string
+}
+
+func NewMetricServer(collectionInterval, port int, metricsEndpointPath string) *MetricServer {
+	return &MetricServer{
+		collectionInterval:  collectionInterval,
+		port:                port,
+		metricsEndpointPath: metricsEndpointPath,
+	}
+}
+
+// Start performs necessary initializations and starts the metric server.
+func (m *MetricServer) Start() error {
+	glog.Infoln("Starting metrics server")
+
+	err := nvml.Init()
```

Comment:
Now we restrict the usage of nvml only within the metric reporting, but i guess in future we should move nvml to the core device plugin logic, i.e. the [ngm](https://github.com/GoogleCloudPlatform/container-engine-accelerators/blob/2651e63a6f7fc3682923219fa94321674c046559/pkg/gpu/nvidia/manager.go#L139)?

Reply:
Completely agree with the future direction. Right now, completely isolating the new functionality behind a kill switch seems to be the safest option.